// @stacks/stacking/dist/constants.esm.js
var StackingErrors;

(function (StackingErrors) {
  StackingErrors[StackingErrors["ERR_STACKING_UNREACHABLE"] = 255] = "ERR_STACKING_UNREACHABLE";
  StackingErrors[StackingErrors["ERR_STACKING_INSUFFICIENT_FUNDS"] = 1] = "ERR_STACKING_INSUFFICIENT_FUNDS";
  StackingErrors[StackingErrors["ERR_STACKING_INVALID_LOCK_PERIOD"] = 2] = "ERR_STACKING_INVALID_LOCK_PERIOD";
  StackingErrors[StackingErrors["ERR_STACKING_ALREADY_STACKED"] = 3] = "ERR_STACKING_ALREADY_STACKED";
  StackingErrors[StackingErrors["ERR_STACKING_NO_SUCH_PRINCIPAL"] = 4] = "ERR_STACKING_NO_SUCH_PRINCIPAL";
  StackingErrors[StackingErrors["ERR_STACKING_EXPIRED"] = 5] = "ERR_STACKING_EXPIRED";
  StackingErrors[StackingErrors["ERR_STACKING_STX_LOCKED"] = 6] = "ERR_STACKING_STX_LOCKED";
  StackingErrors[StackingErrors["ERR_STACKING_PERMISSION_DENIED"] = 9] = "ERR_STACKING_PERMISSION_DENIED";
  StackingErrors[StackingErrors["ERR_STACKING_THRESHOLD_NOT_MET"] = 11] = "ERR_STACKING_THRESHOLD_NOT_MET";
  StackingErrors[StackingErrors["ERR_STACKING_POX_ADDRESS_IN_USE"] = 12] = "ERR_STACKING_POX_ADDRESS_IN_USE";
  StackingErrors[StackingErrors["ERR_STACKING_INVALID_POX_ADDRESS"] = 13] = "ERR_STACKING_INVALID_POX_ADDRESS";
  StackingErrors[StackingErrors["ERR_STACKING_ALREADY_REJECTED"] = 17] = "ERR_STACKING_ALREADY_REJECTED";
  StackingErrors[StackingErrors["ERR_STACKING_INVALID_AMOUNT"] = 18] = "ERR_STACKING_INVALID_AMOUNT";
  StackingErrors[StackingErrors["ERR_NOT_ALLOWED"] = 19] = "ERR_NOT_ALLOWED";
  StackingErrors[StackingErrors["ERR_STACKING_ALREADY_DELEGATED"] = 20] = "ERR_STACKING_ALREADY_DELEGATED";
  StackingErrors[StackingErrors["ERR_DELEGATION_EXPIRES_DURING_LOCK"] = 21] = "ERR_DELEGATION_EXPIRES_DURING_LOCK";
  StackingErrors[StackingErrors["ERR_DELEGATION_TOO_MUCH_LOCKED"] = 22] = "ERR_DELEGATION_TOO_MUCH_LOCKED";
  StackingErrors[StackingErrors["ERR_DELEGATION_POX_ADDR_REQUIRED"] = 23] = "ERR_DELEGATION_POX_ADDR_REQUIRED";
  StackingErrors[StackingErrors["ERR_INVALID_START_BURN_HEIGHT"] = 24] = "ERR_INVALID_START_BURN_HEIGHT";
})(StackingErrors || (StackingErrors = {}));

export { StackingErrors };
